pkg/build/api/helpers.go

```diff
@@ -75,3 +78,58 @@ func HasTriggerType(triggerType BuildTriggerType, bc *BuildConfig) bool {
 	matches := FindTriggerPolicy(triggerType, bc)
 	return len(matches) > 0
 }
+
+// RecordStageAndStepInfo records details about each build stage and step
+func RecordStageAndStepInfo(stages []StageInfo, stageName StageName, stepName StepName, startTime unversioned.Time, endTime unversioned.Time) []StageInfo {
+	var stepDuration int64
+
+	// Make sure that the stages slice is initialized
+	if len(stages) == 0 {
+		stages = make([]StageInfo, 0)
+	}
+
+	// If the stage already exists in the slice, update the endTime and DurationMilliseconds, and append the new step.
+	for stageKey, stageVal := range stages {
+		if stageVal.Name == stageName {
+			stages[stageKey].DurationMilliseconds = endTime.Rfc3339Copy().Time.Sub(stages[stageKey].StartTime.Rfc3339Copy().Time).Nanoseconds() / int64(time.Millisecond)
+			if len(stages[stageKey].Steps) == 0 {
+				stages[stageKey].Steps = make([]StepInfo, 0)
+			}
+			stepDuration = endTime.Rfc3339Copy().Time.Sub(startTime.Rfc3339Copy().Time).Nanoseconds() / int64(time.Millisecond)
+			stages[stageKey].Steps = append(stages[stageKey].Steps, StepInfo{
+				Name:                 stepName,
+				StartTime:            startTime,
+				DurationMilliseconds: stepDuration,
+			})
+			return stages
+		}
+	}
+
+	// If the stageName does not exist, add it to the slice along with the new step.
+	steps := make([]StepInfo, 0)
+	stepDuration = endTime.Rfc3339Copy().Time.Sub(startTime.Rfc3339Copy().Time).Nanoseconds() / int64(time.Millisecond)
+	steps = append(steps, StepInfo{
+		Name:                 stepName,
+		StartTime:            startTime,
+		DurationMilliseconds: stepDuration,
+	})
+	stages = append(stages, StageInfo{
+		Name:                 stageName,
+		StartTime:            startTime,
+		DurationMilliseconds: stepDuration,
+		Steps:                steps,
+	})
+	return stages
+}
+
+// MergeStageAndStepInfo merges one Stages slice into another.
+func MergeStageAndStepInfo(stages []StageInfo, stagesToMerge []StageInfo) ([]StageInfo, int64) {
```

Comment:
The way this works is that Stages are like categories, and Steps are like items in those categories.  The first Step that is added sets the start time for the Stage, and each step that is added extends the duration to that Step's stop time.  So the question would seem to be, what if a second Step with the same name is added?  Should we replace the first one? Go ahead and add this second Step to the Stage? or throw an error?

We shouldn't ever get a duplicate step, but if we did, it would not hurt anything, it would just dump it into the Step slice under the correct Stage and keep on chugging along.